tools/shadowsim/src/shadowsim/shadow_shell.py

```diff
@@ -0,0 +1,179 @@
+                        args=[
+                            "-c",
+                            wrapper_script,
+                        ],
+                    )
+                ],
+            )
+        },
+    )
+    config_path = tmpdir.joinpath("shadow.yaml")
+    config_path.write_text(yaml.safe_dump(config))
+
+    data_dir = tmpdir.joinpath("shadow.data")
+    shadow_args = [
+        f"--data-directory={data_dir}",
+        "--progress=false",
+        str(config_path),
+    ]
+
+    shadow_log_path = tmpdir.joinpath("shadow.log")
+    with shadow_log_path.open("w") as shadow_log:
+        shadow_ps = subprocess.Popen(
+            ["shadow"] + shadow_args,
+            stdout=shadow_log,
+            stderr=subprocess.PIPE,
+        )
+        shadow_stdout = None
+        shadow_exited = False
+        while True:
+            # Try opening the simulated process's stdout if we haven't
+            # successfully done so yet.
+            if shadow_stdout is None:
+                try:
+                    shadow_stdout = open(
+                        str(data_dir.joinpath("hosts/host/bash.1000.stdout")), "rb"
+                    )
+                except FileNotFoundError:
+                    # Not created yet, presumably
+                    pass
+
+            # Pump data from sim stdout to our stdout
+            data = None
+            if shadow_stdout is not None:
+                data = shadow_stdout.read()
+            if data:
+                stdout.write(data)
+                stdout.flush()
+
+            if not data and shadow_exited:
+                # Done
+                break
+
+            if not data:
+                # Wait a bit for data to become ready, or shadow to exit.  We
+                # might be able to do something with inotify to be notified
+                # about the stdout file being created or having data ready, but
+                # it seems more complex than it's worth.
+                try:
+                    shadow_ps.wait(1)
+                    # Mark as exited, but loop again in case more output has
+                    # been written.
+                    shadow_exited = True
+                except subprocess.TimeoutExpired:
+                    pass
+    if preserve:
+        print(f"{progname}: Preserving tmpdir {tmpdir}", file=stderr)
+    else:
+        shutil.rmtree(tmpdir)
+    # if shadow failed, dump its stderr
+    if shadow_ps.returncode:
+        assert shadow_ps.stderr is not None
+        shadow_stderr = shadow_ps.stderr.read().decode()
+        stderr.write(textwrap.indent(shadow_stderr, "shadow: "))
+
+    return shadow_ps.returncode
+
+
+def __main__() -> None:
+    """Raw main, suitable for use with `project.scripts` in `pyproject.toml`"""
+
+    PROGNAME: Final[str] = "shadow-shell"
+
+    parser = argparse.ArgumentParser(prog=PROGNAME)
+    parser.add_argument("-p", "--preserve", action=argparse.BooleanOptionalAction)
+    parser.add_argument("args", nargs="+")
```

Comment:
It might be nice to document functionality. E.g., you can use the `help="blah blah" ` argument to the `add_argument()` func, and maybe also to the `ArgumentParser()` constructor.

Reply:
Done